src/linalg/pivot.jl

```diff
@@ -1,82 +1,36 @@
-right hand side in a left circular shift.
-"""
-function statscholesky(xtx::Symmetric{T}, tol::Real = -1) where {T<:AbstractFloat}
-    n = size(xtx, 2)
-    chpiv = cholesky(xtx, Val(true), tol = T(tol), check = false)
-    chunp = cholesky(xtx, check = false);
-
-    piv = LinearAlgebra.BlasInt.(collect(1:n))
-    r = chpiv.rank
-
-    if r < n
-
-        k = chunp.info
-        if k > r
-            @warn """Fixed-effects matrix may be ill conditioned.
-                     Left-circular shift may fail."""
-        end
-
-        nleft = n
-        while nleft > r
-            # the 0 lowerbound is for MKL compatibility
-            # this arises when the unpivoted Cholesky succeeds but the pivoted
-            # Chokesky estimates less than full rank (see #367)
-            if 0 < k < nleft
-                piv = piv[[1:k-1; k+1:n; k]]
-                chunp = cholesky!(Symmetric(xtx[piv, piv]), check = false)
-            end
-            k = chunp.info
-            nleft -= 1
-        end
-
-        for j = (r+1):n   # an MKL <-> OpenBLAS difference
-            for i = (r+1):j
-                chunp.factors[i, j] = zero(T)
-            end
-        end
+    statsrank(x::Matrix{T}, ranktol::Real=1e-8) where {T<:AbstractFloat}
 
-    end
+Return the numerical column rank and a pivot vector.
 
-    CholeskyPivoted(chunp.factors, chunp.uplo, piv, r, tol, chpiv.info)
-end
+The rank is determined from the absolute values of the diagonal of R from
+a pivoted QR decomposition, relative to the first (and, hence, largest) 
+element of this vector.
 
+In the full-rank case the pivot vector is `collect(axes(x, 2))`.
 """
-    statsqr(x::Matrix{T}, ranktol::Real=1e-8) where {T<:AbstractFloat}
-Return a `QRPivoted` object created from `x` where the pivoting scheme
-retains the original order unless singularity is detected.
-"""
-function statsqr(x::Matrix{T}; ranktol=1e-8) where {T<:AbstractFloat}
-    n = size(x, 2)
+function statsrank(x::AbstractMatrix{T}; ranktol=1e-8) where {T<:AbstractFloat}
+    m, n = size(x)
 
     qrpiv = qr(x, Val(true))
-    rank =  searchsortedlast(abs.(diag(qrpiv.R)), ranktol, rev=true);
-    if rank < n
-        piv = qrpiv.p
-    else
-        piv = LinearAlgebra.BlasInt.(collect(1:n))
+    dvec = abs.(diag(qrpiv.R))
+    fdv = first(dvec)
+    if (last(dvec) / fdv) > ranktol
+        return (rank = n, piv = collect(axes(x, 2)))
     end
-
-    if rank < n && piv[1] ≠ 1
-        # make sure the first column isn't moved
-        # this is usually the intercept and it's desirable to avoid
-        # pivoting the intercept out. note that this inflation may
-        # not always be enough, but if it isn't, then you probably
-        # have other problems
-
-        x = deepcopy(x) # this is horrible
-        x[:, 1] .*= 1e6
+    rank = searchsortedlast(dvec ./ fdv, ranktol, rev=true)
+    @assert rank < n
+    piv = qrpiv.p
+    v1 = first(eachcol(x))
+    if all(isone, v1) && first(piv) ≠ 1
```

Comment:
I like that this only prevents moving the first column if it looks like an intercept